anomaly_detection/db/models.py

```diff
@@ -13,13 +13,162 @@
+        #                 raises NoneType exception if there is no running
+        #                 transaction and rollback is called. As long as
+        #                 sqlalchemy has this bug we have to create transaction
+        #                 explicitly.
+        with session.begin(subtransactions=True):
+            session.add(self)
+            session.flush()
+
+    def __setitem__(self, key, value):
+        setattr(self, key, value)
+
+    def __getitem__(self, key):
+        return getattr(self, key)
+
+    def __contains__(self, key):
+        # Don't use hasattr() because hasattr() catches any exception, not only
+        # AttributeError. We want to passthrough SQLAlchemy exceptions
+        # (ex: sqlalchemy.orm.exc.DetachedInstanceError).
+        try:
+            getattr(self, key)
+        except AttributeError:
+            return False
+        else:
+            return True
+
+    def get(self, key, default=None):
+        return getattr(self, key, default)
+
+    @property
+    def _extra_keys(self):
+        """Specifies custom fields
+
+        Subclasses can override this property to return a list
+        of custom fields that should be included in their dict
+        representation.
+
+        For reference check tests/db/sqlalchemy/test_models.py
+        """
+        return []
+
+    def __iter__(self):
+        columns = list(dict(object_mapper(self).columns).keys())
+        # NOTE(russellb): Allow models to specify other keys that can be looked
+        # up, beyond the actual db columns.  An example would be the 'name'
+        # property for an Instance.
+        columns.extend(self._extra_keys)
+
+        return ModelIterator(self, iter(columns))
+
+    def update(self, values):
+        """Make the model object behave like a dict."""
+        for k, v in values.items():
+            setattr(self, k, v)
+
+    def _as_dict(self):
+        """Make the model object behave like a dict.
+
+        Includes attributes from joins.
+        """
+        local = dict((key, value) for key, value in self)
+        joined = dict([(k, v) for k, v in self.__dict__.items()
+                      if not k[0] == '_'])
+        local.update(joined)
+        return local
+
+    def iteritems(self):
+        """Make the model object behave like a dict."""
+        return self._as_dict().items()
+
+    def items(self):
+        """Make the model object behave like a dict."""
+        return self._as_dict().items()
+
+    def keys(self):
+        """Make the model object behave like a dict."""
+        return [key for key, value in self.iteritems()]
+
+
+class TimestampMixin(object):
+    created_at = Column(DateTime, default=lambda:  datetime.datetime.utcnow())
+    updated_at = Column(DateTime, onupdate=lambda: datetime.datetime.utcnow())
+
+
+Base = declarative_base()
+
+
+class AnomalyDetectionBase(ModelBase, TimestampMixin):
+    """Base class for Manila Models."""
+    __table_args__ = {'mysql_engine': 'InnoDB'}
```

Comment:
We are going to use mongoDB.  Why are you using mysql here?

Also we should make the database a pluggable framework to allow different database type to be used in the future.

Reply:
done